src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -119,40 +119,42 @@ private void OnUnexpectedExceptionThrown(Exception unexpectedException)
             HostGroup hostGroup,
             CancellationToken cancellationToken)
         {
-            Exception unexpectedException;
+            const string LogMessage = "Unexpected exception from HubClient";
 
             var descriptor = new ServiceDescriptor(serviceName) { HostGroup = hostGroup };
             try
             {
                 return await client.RequestServiceAsync(descriptor, cancellationToken).ConfigureAwait(false);
             }
-            catch (Exception e) when (e is ConnectionLostException || e is RemoteInvocationException re && re.InnerException is InvalidOperationException)
+            catch (Exception e) when (ReportNonFatalWatson(e, cancellationToken))
             {
-                // ServiceHub may throw these exceptions if it is called after VS started to shut down.
-                // Cancel the operation and do not report an error in these cases.
+                // TODO: Once https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1040692.
+                // ServiceHub may throw non-cancellation exceptions if it is called after VS started to shut down,
+                // even if our cancellation token is signaled. Cancel the operation and do not report an error in these cases.
                 // 
-                // TODO: Once https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1040692 is implemented,
-                // i.e. the ServiceHub does not throw these exceptions when cancellation token is signaled,
+                // If ServiceHub did not throw non-cancellation exceptions when cancellation token is signaled,
                 // we can assume that these exceptions indicate a failure and should be reported to the user.
-                // That's because we expect the cancellation to be requested on the token before HubClient gets disposed.
                 cancellationToken.ThrowIfCancellationRequested();
 
-                unexpectedException = e;
-            }
-            catch (Exception e) when (!(e is OperationCanceledException))
-            {
-                unexpectedException = e;
+                client.Logger.TraceEvent(TraceEventType.Error, 1, $"{LogMessage}: {e}");
+
+                RemoteHostCrashInfoBar.ShowInfoBar(workspace, e);
+
+                // TODO: Propagate the original exception (see https://github.com/dotnet/roslyn/issues/40476)
+                throw new SoftCrashException(LogMessage, e, cancellationToken);
             }
```

Comment:
more cases of not understanding why this is hte right exception pattern.

Reply:
Yeah, so this is complicated. 

As the comment says, `HubClient` sometimes (on shutdown) throws random exceptions even if the `cancellationToken` is signaled. They are fixing that (the linked devdiv bug). Meanwhile we need to treat any exception as cancellation if the token is signaled. Hence https://github.com/dotnet/roslyn/pull/40395/files#diff-04c9f43587bd90007aa056edc7dd7fa7R137 and https://github.com/dotnet/roslyn/pull/40395/files#diff-04c9f43587bd90007aa056edc7dd7fa7R151. Now if the token is not signaled then the exception is not expected. So we show info bar telling the user that we failed and they need to restart VS because we are in bad state.

We throw `SoftCrashException` for now since not all services are hardened against unexpected exceptions, although they should be. This is tracked by #40476. Once we harden the code we can just let the original exception propagate.